### PR TITLE
#187419169 user can login via email and password 

### PR DESCRIPTION
**Pull request by niyobertin**
**What does this PR do?**

Validation of user login data

**Description of Task to be completed?**

The user data from request.body are validated using Joi

**How should this be manually tested?**

Clone this repository , cd into it and run:

npm run dev to start server . while server running, use postman and send
1.POST request on locahost:PORT/api/v1/users/login by providing email and password.
it will return jwt as result of successful login.
in case of failure it will return error based on failing field.

**What are the relevant pivotal tracker/Trello stories?**

[#187419169](https://www.pivotaltracker.com/n/projects/2701065)

### CODE REVIEW

**Comment by teerenzo** [comment]
squash your commits to be one commit

File: src/moddewares/joiValifdation.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const unknownFields = Object.keys(req.body).filter(field => !allowedFields.includes(field));
+    if (unknownFields.length > 0) {
+        return res.status(406).json({
+            status: 406,
+            message: `Unknown fields: ${unknownFields.join(", ")}`
+        });
+    }else{
+        next();
+    }
+}
```
Comment (teerenzo) [inline src/moddewares/joiValifdation.ts]: remove this if you are not using it

**Review by teerenzo — CHANGES_REQUESTED** [review]

**Review by Jovz19200 — APPROVED** [review]
Awesome !!!

File: src/helpers/validation.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+       return res.status(406)
+        .json({
+            status:406,
+            mesage:`Error in User Data : ${error.message}`
+        })
+    }
+    const allowedFields = Object.keys(data.describe().keys);
+    const unknownFields = Object.keys(req.body).filter(field => !allowedFields.includes(field));
+    if (unknownFields.length > 0) {
+        return res.status(406).json({
```
Comment (soleil00) [inline src/helpers/validation.ts]: wrong status u should use 400 here 406 is used for unacceptable which is related to content type but u used it for field validations

File: src/middlewares/joiValidation.ts
Comment (soleil00) [inline src/middlewares/joiValidation.ts:1]: creating joiValidation in helper and creating validation in middleware on my point of view its unnecessary u can either create middleware that handle both loginValidation schema u can move it into validation in middleware or mode down validation in middleware into helper

**Review by soleil00 — APPROVED** [review]
despit comment ur fix wor as expected , u too u may consider refacto in near the future

File: __test__/user.test.ts [hunk 3]
```diff
@@ -94,21 +94,6 @@ describe("Testing user Routes", () => {
...
-    };
-
-    const response = await request(app).post("/api/v1/users/login").send({
-      email: dummySeller.email,
-      password: dummySeller.password,
-    });
-
-    expect(response.body.message).toBe("Verification link has been sent to your email. Please verify it to continue");
-  }, 20000);
-
```
Comment (teerenzo) [inline __test__/user.test.ts]: why did you remove this

**Review by teerenzo — CHANGES_REQUESTED** [review]

**Comment by teerenzo** [comment]
I can't see where you have validated your fields on login @niyobertin please pay attention to details and test before raising any PR

File: src/routes/userRoutes.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import { Router } from "express";
 import { fetchAllUsers, createUserController, userLogin, updatePassword } from "../controllers/userControllers";
-import { emailValidation, validateSchema } from "../middleware/validator";
+import { emailValidation, validateSchema } from "../middlewares/validator";
 import signUpSchema from "../schemas/signUpSchema";
 import { isLoggedIn } from "../middlewares/isLoggedIn";
 import { passwordUpdateSchema } from "../schemas/passwordUpdate";
```
Comment (teerenzo) [inline src/routes/userRoutes.ts]: where is validation of login

**Review by teerenzo — CHANGES_REQUESTED** [review]

File: src/routes/userRoutes.ts [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
 import { otpVerification } from "../controllers/2faControllers";
 
 const userRoutes = Router();
 
 userRoutes.get("/", fetchAllUsers);
-userRoutes.post("/login", userLogin);
+userRoutes.post("/login",validateSchema(logInSchema), userLogin);
 userRoutes.post("/register", emailValidation, validateSchema(signUpSchema), createUserController);
```
Comment (teerenzo) [inline src/routes/userRoutes.ts:15]: what about email

**Review by teerenzo — CHANGES_REQUESTED** [review]

**Pull request closed without merging**